api/envoy/config/cluster/v3/outlier_detection.proto

```diff
@@ -197,47 +177,7 @@ message OutlierDetection {
   // Defaults to true.
   google.protobuf.BoolValue successful_active_health_check_uneject_host = 23;
 
-  // Monitor which counts consecutive errors.
-  // If number of consecutive errors exceeds the threshold, monitor will report that the host
-  // is unhealthy.
-  // [#not-implemented-hide:]
-  message ConsecutiveErrors {
-    // Monitor name.
-    string name = 1;
-    // The number of consecutive errors before ejection occurs.
-    google.protobuf.UInt32Value threshold = 2 [(validate.rules).uint32 = {lte: 100}];
-    // The % chance that a host is actually ejected. Defaults to 100.
-    google.protobuf.UInt32Value enforcing = 3 [(validate.rules).uint32 = {lte: 100}];
-    // List of error buckets.
-    repeated ErrorBucket error_buckets = 4;
-  }
-
-  // Monitor which counts percentage of failed requests.
-  // Percentage of failed requests is calculated at each interval. If the percentage of failed requests is larger than
-  // defined threshold, the monitor will report that the host is unhealthy.
-  // [#not-implemented-hide:]
-  message FailurePercentage {
-    // Monitor name.
-    string name = 1;
-    // The percentage of failed requests required to mark the host as unhealthy. Defaults to 85.
-    google.protobuf.UInt32Value threshold = 2 [(validate.rules).uint32 = {lte: 100}];
-    // The % chance that a host is actually ejected. Defaults to 100.
-    google.protobuf.UInt32Value enforcing = 3 [(validate.rules).uint32 = {lte: 100}];
-    // List of error buckets.
-    repeated ErrorBucket error_buckets = 4;
-  }
-
-
-  // Set of passive monitors.
-  // [#not-implemented-hide:]
-  message Monitors {
-    // Set of monitors counting consecutive errors.
-    repeated ConsecutiveErrors consecutive_errors = 1;
-    // Set of failure percentage monitors.
-    repeated FailurePercentage failure_percantage = 2;
-  }
-
   // Set of host's passive monitors.
   // [#not-implemented-hide:]
-  Monitors monitors = 24;
+  repeated Monitor monitors = 24;
```

Comment:
Hi, what role will the `monitor` take? Will it handle all the logic of outlier detection or just determine if an error is throwed?

And is the `repeated` necessary? Will all monitors work at same time or just first valid monitor will be used?

If all `monitors` work at same time, it would be more complex to handle the output of different monitor. I prefer only use `TypedExtensionConfig monitor = 24;` here because single monitor is enough for absolute majority of users. If some one wants to composite multiple monitor, he can implement it in an extension.

Reply:
It's worth noting that the existing mechanism for configuring outlier detection does allow configuring both failure percentage and success rate at the same time.  If the goal here is to move that existing configuration into an extension, then it seems like we should make it possible to configure multiple ejection algorithms.

Reply:
> Hi, what role will the monitor take? Will it handle all the logic of outlier detection or just determine if an error is throwed?

A monitor will consume results of interaction with upstream entity and run logic specific to a monitor. For example, count consecutive errors or standard deviation or frequency of errors.

> And is the repeated necessary? Will all monitors work at same time or just first valid monitor will be used?

The result of a transaction will go to all monitors.

> If all monitors work at same time, it would be more complex to handle the output of different monitor. 

You are right that most users will use a single monitor. But, as @markdroth pointed out, there are already many monitors, so `repeated Monitor` is a natural extension. Also, I plan to add a monitor which measures response time and eject slow servers. The combination of one monitor which counts errors and another monitor which measures response time starts to make sense.

I think that the idea of multiple monitors was always there. See this comment: https://github.com/envoyproxy/envoy/blob/release/v1.28/source/common/upstream/outlier_detection_impl.h#L369-L371. But for some reason, instead of adding new, separate monitors with different logic, all algorithms were added to a single monitor. With this new API, I hope to change it and maybe, if there is an agreement, deprecate old ways of configuring outlier detector.

Reply:
sgtm if we intend to use the monitor replace whole previous implementation.